BrowserKit/Sources/Common/Theming/DefaultThemeManager.swift

```diff
@@ -83,132 +78,102 @@ public final class DefaultThemeManager: ThemeManager, Notifiable {
+            applyThemeChanges(for: $0, using: determineThemeType(for: $0))
         }
-
-        // After updating all windows, notify (once). We send the UUID of the window for
-        // which the change originated though more than 1 window may ultimately update its UI.
-        notifyCurrentThemeDidChange(for: window)
     }
 
-    public func reloadTheme(for window: WindowUUID) {
-        updateCurrentTheme(to: fetchSavedThemeType(for: window), for: window)
+    // MARK: - Manual theme functions
+    public func setManualTheme(to newTheme: ThemeType) {
+        updateSavedTheme(to: newTheme)
+        applyThemeUpdatesToWindows()
     }
 
-    public func systemThemeChanged() {
-        allWindowUUIDs.forEach { uuid in
-            // Ignore if:
-            // the system theme is off
-            // OR night mode is on
-            // OR private mode is on
-            guard systemThemeIsOn,
-                  !nightModeIsOn,
-                  !privateModeIsOn(for: uuid)
-            else { return }
-
-            changeCurrentTheme(getSystemThemeType(), for: uuid)
-        }
+    public func getUserManualTheme() -> ThemeType {
+        guard let savedThemeDescription = userDefaults.string(forKey: ThemeKeys.themeName),
+              let savedTheme = ThemeType(rawValue: savedThemeDescription)
+        else { return getThemeTypeBasedOnSystem() }
+
+        return savedTheme
     }
 
+    // MARK: - System theme functions
     public func setSystemTheme(isOn: Bool) {
         userDefaults.set(isOn, forKey: ThemeKeys.systemThemeIsOn)
+        applyThemeUpdatesToWindows()
+    }
 
-        if isOn {
-            systemThemeChanged()
-        } else if automaticBrightnessIsOn {
-            updateThemeBasedOnBrightness()
-        } else {
-            allWindowUUIDs.forEach { reloadTheme(for: $0) }
-        }
+    private func getThemeTypeBasedOnSystem() -> ThemeType {
+        return UIScreen.main.traitCollection.userInterfaceStyle == .dark ? ThemeType.dark : ThemeType.light
     }
 
+    // MARK: - Private theme functions
     public func setPrivateTheme(isOn: Bool, for window: WindowUUID) {
-        let currentSetting = getPrivateThemeIsOn(for: window)
-        guard currentSetting != isOn else { return }
+        guard getPrivateThemeIsOn(for: window) != isOn else { return }
 
         var settings: KeyedPrivateModeFlags
         = userDefaults.object(forKey: ThemeKeys.PrivateMode.byWindowUUID) as? KeyedPrivateModeFlags ?? [:]
 
         settings[window.uuidString] = NSNumber(value: isOn)
         userDefaults.set(settings, forKey: ThemeKeys.PrivateMode.byWindowUUID)
 
-        updateCurrentTheme(to: fetchSavedThemeType(for: window), for: window)
+        applyThemeChanges(for: window, using: determineThemeType(for: window))
     }
 
     public func getPrivateThemeIsOn(for window: WindowUUID) -> Bool {
         let settings = userDefaults.object(forKey: ThemeKeys.PrivateMode.byWindowUUID) as? KeyedPrivateModeFlags
```

Comment:
Doesn't need to block the PR but we probably don't need to store this in `UserDefaults` any longer, and changing it might also fix this issue where the windows are rendering briefly in Private theme incorrectly during launch:

https://github.com/mozilla-mobile/firefox-ios/assets/145381717/a2da2c06-41f9-4eb2-8eb8-396445abec4c